jaxrs/WFLY-13530_expose_JAXRS_to_gRPC.adoc

```diff
@@ -0,0 +1,133 @@
+= Expose JAX-RS Resources to gRPC clients
+:author:            Ron Sigal
+:email:             rsigal@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+:issue-base-url:    https://issues.jboss.org/browse
+
+== Overview
+
+In the five years since gRPC was introduced, it has become a popular invocation
+and streaming framework, used by major corporations. Among its benefits are
+
+* speed
+* language independence
+
+With this feature request we propose to expose JAX-RS services as gRPC services,
+accessible by gRPC clients.
+
+Given a gRPC service definition like
+----
+service Greeter {
+  rpc SayHello (HelloRequest) returns (HelloReply) {}
+}
+----
+
+the protobuf java compiler will create a class GreeterGrpc, and a service can be
+created by subclassing. We envision making it as easy as something like this
+to forward a gRPC request to JAX-RS:
+
+----
+
+static class GreeterImpl extends GreeterGrpc.GreeterImplBase {
+
+   public void sayHello(HelloRequest req, StreamObserver<HelloReply> responseObserver) {
+      JAXRSForwarderBuilder builder = new JAXRSForwarderBuilder();
```

Comment:
Is this an illustration of how the service would be generated or would the developer be expected to implement the service to forward to JAX-RS?

Reply:
In general, I think of it as an illustration. There's got to be a way of translating the gRPC invocation into something suitable for the target JAX-RS resource. Now, if the JAX-RS resource is written to match the gRPC service definition, then the translation could be automated. In fact, I imagine a simple default JAXRSForwarder that does that. But if the scenario involves a gRPC client for an existing JAX-RS resource, then the translation would be more complex. Now that you mention it, it might be interesting to think about automating the process, but it's not clear to me how feasible that would be.

For now, though, I'm just thinking about a way to make the translation process as easy as possible to define.